eclair-core/src/main/scala/fr/acinq/eclair/router/RouteCalculation.scala

```diff
@@ -219,4 +222,137 @@ object RouteCalculation {
+                         pendingHtlcs: Seq[Route] = Nil,
+                         routeParams: RouteParams,
+                         currentBlockHeight: Long): Try[Seq[Route]] = Try {
+    val result = findMultiPartRouteInternal(g, localNodeId, targetNodeId, amount, maxFee, extraEdges, ignoredEdges, ignoredVertices, pendingHtlcs, routeParams, currentBlockHeight) match {
+      case Right(routes) => Right(routes)
+      case Left(RouteNotFound) if routeParams.randomize =>
+        // If we couldn't find a randomized solution, fallback to a deterministic one.
+        findMultiPartRouteInternal(g, localNodeId, targetNodeId, amount, maxFee, extraEdges, ignoredEdges, ignoredVertices, pendingHtlcs, routeParams.copy(randomize = false), currentBlockHeight)
+      case Left(ex) => Left(ex)
+    }
+    result match {
+      case Right(routes) => routes
+      case Left(ex) => return Failure(ex)
+    }
+  }
+
+  private def findMultiPartRouteInternal(g: DirectedGraph,
+                                         localNodeId: PublicKey,
+                                         targetNodeId: PublicKey,
+                                         amount: MilliSatoshi,
+                                         maxFee: MilliSatoshi,
+                                         extraEdges: Set[GraphEdge] = Set.empty,
+                                         ignoredEdges: Set[ChannelDesc] = Set.empty,
+                                         ignoredVertices: Set[PublicKey] = Set.empty,
+                                         pendingHtlcs: Seq[Route] = Nil,
+                                         routeParams: RouteParams,
+                                         currentBlockHeight: Long): Either[RouterException, Seq[Route]] = {
+    // We use Yen's k-shortest paths to find many paths for chunks of the total amount.
+    val numRoutes = {
+      val directChannelsCount = g.getEdgesBetween(localNodeId, targetNodeId).length
+      routeParams.mpp.maxParts.max(directChannelsCount) // if we have direct channels to the target, we can use them all
+    }
+    val routeAmount = routeParams.mpp.minPartAmount.min(amount)
+    findRouteInternal(g, localNodeId, targetNodeId, routeAmount, maxFee, numRoutes, extraEdges, ignoredEdges, ignoredVertices, routeParams, currentBlockHeight) match {
+      case Right(routes) =>
+        // We use these shortest paths to find a set of non-conflicting HTLCs that send the total amount.
+        split(amount, mutable.Queue(routes: _*), initializeUsedCapacity(pendingHtlcs), routeParams) match {
+          case Right(routes) if validateMultiPartRoute(amount, maxFee, routes) => Right(routes)
+          case _ => Left(RouteNotFound)
+        }
+      case Left(ex) => Left(ex)
+    }
+  }
+
+  private def split(amount: MilliSatoshi, paths: mutable.Queue[Graph.WeightedPath], usedCapacity: mutable.Map[ShortChannelId, MilliSatoshi], routeParams: RouteParams): Either[RouterException, Seq[Route]] = {
```

Comment:
Good point, worth investigating! I don't think so because we'll have a small number of `paths` (less than 10) so it should be well-bounded. I'll take a stab at making this `tailrec` and if the code is still simple enough to read, that should remove the concern.

Reply:
I made it tailrec in https://github.com/ACINQ/eclair/pull/1427/commits/a78baed7df824a68f477d73c8c9c38743f87cce3, this is always worth it!